Explain why Elasticsearch doesn't support incremental resharding.

jpountz opened the pull request:
I have seen this question a couple times already, most recently at
https://twitter.com/dimosr7/status/973872744965332993

I tried to keep the explanation as simple as I could, which is not always easy
as this is a matter of trade-offs.

elasticmachine commented:
Pinging @elastic/es-core-infra

DaveCTurner commented on docs/reference/indices/split-index.asciidoc:
```diff
@@ -47,6 +49,27 @@ Splitting works as follows:
+Going from `N` shards to `N+1` shards, aka. incremental resharding is indeed a
+feature that is supported by many key-value stores, which leverage the fact that
+consistent hashing only requires `1/N`-th of the keys to be relocated. However
+Elasticsearch's unit of storage, shards, are Lucene indices. Unfortunately,
```
I think we can leave `Unfortunately` out. Maybe something like `Because of their search-optimised structure`, or maybe just nothing at all.

nik9000 replied:
I like "because of their search-optimized structure" because it explains why it isn't a thing for us.

DaveCTurner commented on docs/reference/indices/split-index.asciidoc:
```diff
@@ -47,6 +49,27 @@ Splitting works as follows:
+[float]
+=== Why doesn't Elasticsearch support incremental resharding?
+
+Going from `N` shards to `N+1` shards, aka. incremental resharding is indeed a
```
I think there should be a comma after `resharding` to close the parenthesis.

nik9000 commented on docs/reference/indices/split-index.asciidoc:
```diff
@@ -47,6 +49,27 @@ Splitting works as follows:
+
+Moreover, only allowing large increments of the number of shards also has the
+good property of reducing the number of times that an index needs to be split
+during its lifetime.
```
I don't like this paragraph all that much. It comes off as paternalistic. "You don't want this feature anyway." I figure we should just drop the paragraph and let users decide if they want the feature or not. If you feel like we have to keep the paragraph we should rework it so this sentiment comes across as a silver lining.

jpountz replied:
You made a good point. I removed it.

nik9000 commented:
Thanks for writing this!

jpountz commented:
Thanks @DaveCTurner  and @nik9000  for you comments, I addressed your feedback. I had missed @bleskes 's reply - for some reason my Twitter client thought it was not worth displaying!

bleskes commented on docs/reference/indices/split-index.asciidoc:
```diff
@@ -47,6 +49,28 @@ Splitting works as follows:
+
+Going from `N` shards to `N+1` shards, aka. incremental resharding, is indeed a
+feature that is supported by many key-value stores, which leverage the fact that
+consistent hashing only requires `1/N`-th of the keys to be relocated. However
```
This is already great, but I feel it will be even clear if we explain that we can't just open a new empty shard and starting indexing into it - that will just create a hotspot and will result in a complicated id routing logic. This means that we have rebalance existing data. That is difficult as is already explained.

jpountz commented:
@bleskes I pushed a new update, let me know what you think.
